### PR TITLE
Fixes ModelParamTraversal & value-dependent filter

### PR DESCRIPTION
**Pull request by andsteing**
Follow up to #1135

### CODE REVIEW

**Comment by andsteing** [comment]
The fix proposed in #1135 does not work with `ModelParamTraversal` that have a `filter_fn` that filters based on the `value` of a leaf. This PR expands the solution to fake the `.dtype` and `.shape` attributes, while maintaining the `key` and `sub_optimizers` indexing for improved error messaging

File: flax/optim/base.py [hunk 1]
```diff
@@ -479,18 +493,18 @@ def __init__(
   def init_state(self, params):
     sub_states = []
     seen = {}
+    dummies = traverse_util.unflatten_dict({
```
Comment (jheek) [inline flax/optim/base.py:496]: Is it intentional to still use flatten_dict here? It's nice for the error message but it still breaks non-dict optimizer targets this way.

Comment (andsteing) [reply flax/optim/base.py]: Oh - I was only thinking of `ModelParamTraversal`...

PTAL, should work with arbitrary traversals now (incl. useful error message - assuming containers know how to `__str__`).

**Comment by andsteing** [comment]
Note that`MultiOptimizer` might change some more soon : #1177

**Review by jheek — APPROVED** [review]
Looks good!

**Pull request merged**